### PR TITLE
[quant][graphmode] Handling ops doesn't require observation in insertObservers

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33650 [not4land] debugging
* #33533 [not4land] Temporary workaround to get quantized ops
* #34414 [quant][graphmode][refactor] insertObservers for Block
* #34411 [quant][graphmode][fix] Insert dequantize before use node
* #34349 [quant][graphmode][fix] preserve the type of original value when inserting dequant node
* #34348 [quant] Add dequantize.tensors
* #34347 [quant][graphmode] Add quantization support for aten::dropout
* #34346 [quant][graphmode] Add quantization support for aten::cat
* #34345 [quant][graphmode] Add prim::ListConstruct to general op handling
* #34315 [quant][graphmode] insert quant/dequant work for duplicated debugName
* #34314 [quant][graphmode][refactor] Change QParamMap to QParamVector
* #33895 [quant][graphmode] insert prepack/unpack for aten::linear
* #33854 [quant][graphmode] Quantization pattern for aten::linear
* #33532 [quant][graphmode] Add quantization pattern for quantized::add_relu
* #33279 [quant][graphmode] Add quantized conv2d-relu fusion pattern
* #33927 [quant][graphmode] Add Finalize function that inlines graph and produce quantized ops
* #33853 [quant][graphmode] Swap quantized functional linear with aten::linear
* **#33481 [quant][graphmode] Handling ops doesn't require observation in insertObservers**

Summary:
We have to propagate observed property of values through ops like max_pool2d, flatten and
avoid inserting duplicated observers.
For example:
```
x1 = self.conv(x)
x2 = maxpool(x1)
x3 = self.conv(x2)
```
If x1 is observed, we should propagate this information through maxpool and
we should consider x2 as observed as well.

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D20261897](https://our.internmc.facebook.com/intern/diff/D20261897)

### CODE REVIEW

File: torch/csrc/jit/passes/quantization.cpp [hunk 1]
```diff
@@ -776,6 +791,28 @@ void InsertObserversHelper::addIntermediateValuesToSkipObserver(
...
 
+void InsertObserversHelper::fillPassThroughValueMap(const std::shared_ptr<Graph>& graph) {
+  std::stack<Block*> blocks_to_visit;
+  blocks_to_visit.push(graph->block());
+  while (!blocks_to_visit.empty()) {
+    Block* b = blocks_to_visit.top();
+    blocks_to_visit.pop();
+    for (Node* n : b->nodes()) {
+      auto input_indexes = getGeneralOpTensorInputIndexes(n);
+      if (input_indexes.size() > 0) {
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp]: oh right, will remove. This is copied from SwapDeQuant function in this file

File: torch/csrc/jit/passes/quantization.cpp [hunk 2]
```diff
@@ -481,16 +490,30 @@ class InsertObserversHelper {
...
+  void fillPassThroughValueMap(const std::shared_ptr<Graph>& graph);
+
   const ModuleQConfigMap& module_qconfig_map_;
   // Values we want to skip observing, used to skip values in
   // the middle of the ops that are supposed to be fused, e.g.
   // the output value of conv in the conv - relu pattern
   std::unordered_set<Value*> values_to_skip_;
   std::unordered_set<Graph*> visited_graph_of_observer_map_;
   std::unordered_map<Value*, script::Module> observer_for_value_;
+  std::unordered_map<Value*, Value*> caller_to_callee_;
```
Comment (raghuramank100) [inline torch/csrc/jit/passes/quantization.cpp]: Is this comment correct? This seems like a description of observer_for_value

**Review by raghuramank100 — APPROVED** [review]
Looks good!

File: torch/csrc/jit/passes/quantization.cpp (hunk 2 above)
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization.cpp]: which comment?

**Pull request closed without merging**